### PR TITLE
Limit number of records that can be accumulated by Jet processors

### LINKED ISSUE #18590: Implement basic memory management for Jet SQL
Jet SQL supports operations which potentially consume unbounded amounts of memory. We need to implement at least basic memory management so that the stability of the cluster isn't threatened due to an OOME. For example, cap the amount of memory consumed by one query.

### PR DESCRIPTION
**Pull request by gierlachg**
Fixes #18590

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/jet/core/test/TestProcessorMetaSupplierContext.java [hunk 1]
```diff
@@ -173,6 +173,11 @@ public ProcessingGuarantee processingGuarantee() {
         return processingGuarantee;
     }
 
+    @Override
+    public long maxProcessorAccumulatedRecords() {
+        return Long.MAX_VALUE;
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/core/test/TestProcessorMetaSupplierContext.java:178]: We should have a field here that the user can set. One can then test processors with a desired value.

File: hazelcast/src/main/java/com/hazelcast/jet/impl/memory/MaxProcessorAccumulatedItemsExceededException.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+package com.hazelcast.jet.impl.memory;
+
+import com.hazelcast.jet.config.InstanceConfig;
+
+/**
+ * This exception is thrown when a job exceeds a configurable processor accumulation limit.
+ *
+ * @see InstanceConfig#setMaxProcessorAccumulatedRecords
+ */
+public class MaxProcessorAccumulatedItemsExceededException extends RuntimeException {
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/memory/MaxProcessorAccumulatedItemsExceededException.java:26]: Name suggestion: `AccumulationLimitExceededException`

**Review by viliam-durina — APPROVED** [review]

**Comment by gierlachg** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/jet/impl/processor/GroupP.java [hunk 3]
```diff
@@ -70,12 +74,23 @@ public GroupP(
...
+    }
+
     @Override
     @SuppressWarnings("unchecked")
     protected boolean tryProcess(int ordinal, @Nonnull Object item) {
         Function<Object, ? extends K> keyFn = (Function<Object, ? extends K>) groupKeyFns.get(ordinal);
         K key = keyFn.apply(item);
-        A acc = keyToAcc.computeIfAbsent(key, k -> aggrOp.createFn().get());
+        A acc = keyToAcc.computeIfAbsent(key, k -> {
+            if (keyToAcc.size() == maxEntries) {
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/jet/impl/processor/GroupP.java:88]: Why is this comparing the number of keys? Shouldn't it rather compare number of accumulated values?

Also some aggregation functions keep the items (e.g. collect) and some don't (e.g. counting). Shouldn't these be handled differently?

**Review by frant-hartm — CHANGES_REQUESTED** [review]
Could you explain (e.g. in the commit message) why this does not apply to the streaming aggregations.

Also does it make sense to apply to TransformStatefulP?

File: hazelcast/src/main/java/com/hazelcast/jet/config/InstanceConfig.java [hunk 4]
```diff
@@ -44,13 +45,22 @@
...
+     * The default value of the {@link #setMaxProcessorAccumulatedRecords(long) max processor accumulated records}.
+     */
+    private static final long MAX_PROCESSOR_ACCUMULATED_RECORDS = Long.MAX_VALUE;
+
     private int cooperativeThreadCount = RuntimeAvailableProcessors.get();
     private int flowControlPeriodMs = DEFAULT_FLOW_CONTROL_PERIOD_MS;
     private int backupCount = DEFAULT_BACKUP_COUNT;
     private long scaleUpDelayMillis = SCALE_UP_DELAY_MILLIS_DEFAULT;
     private boolean losslessRestartEnabled;
+    private long maxProcessorAccumulatedRecords = MAX_PROCESSOR_ACCUMULATED_RECORDS;
```
[the remark this answers is not in the document]
Comment (gierlachg) [reply hazelcast/src/main/java/com/hazelcast/jet/config/InstanceConfig.java:63]: I have a follow up PR that adds the possibility to configure it on a job level.

File: hazelcast/src/main/java/com/hazelcast/jet/impl/processor/GroupP.java (hunk 3 above)
Comment (gierlachg) [reply hazelcast/src/main/java/com/hazelcast/jet/impl/processor/GroupP.java:87]: We wanted to make it as simple as possible -> we count just entries, irrelevant of their sizes. To solve all the issues we would need to track each and every object size on a byte level (even single object with i.e. `minBy()` can take a lot of space).
Moreover, from what I have learnt `collecting` aggregations are not that common.

**Comment by gierlachg** [comment]
Regarding streaming aggregations - quoting @viliam-durina:
> You don't want your streaming job to fail due to some arbitrary boundary. The consumption of memory should be more or less stable, and you should monitor it. If it's growing (due to traffic growing over time), the user should spot it early enough.

**Comment by frant-hartm** [comment]
While I agree with the monitoring part, a OOM will likely kill whole member, maybe even cluster. This is especially bad if there are multiple jobs running/applications connected to it, it would be better for the single streaming job to die.

**Review by frant-hartm — APPROVED** [review]

**Comment by gierlachg** [comment]
run-lab-run

**Comment by viliam-durina** [comment]
> While I agree with the monitoring part, a OOM will likely kill whole member, maybe even cluster. This is especially bad if there are multiple jobs running/applications connected to it, it would be better for the single streaming job to die.

For batch jobs, the memory usage grows all the time, until the end of job. For streaming jobs, it should be stable (after an initial growth) and can even go down if the traffic goes down. Batch jobs tend to be ad-hoc, especially through sql. A simple mistake in the sql can cause it consume a lot of resources, e.g. grouping by a wrong key, or just adding `DISTINCT`.

I think we should also add tests for SQL jobs that they fail if the count is exceeded. We might change the backing processor and forget to implement the limit.

**Pull request merged**